File: Sample Apps/Objc/SampleApp/MessagingViewController.m
//
//  ViewController.m
//  SampleApp
//
//  Created by Nimrod Shai on 2/11/16.
//  Copyright © 2016 LivePerson. All rights reserved.
//

#import "MessagingViewController.h"
#import <LPMessagingSDK/LPMessagingSDK.h>


#define WINDOW_SWITCH @"windowSwitch"
#define AUTHENTICATION_SWITCH @"authenticationSwitch"


@interface MessagingViewController ()<LPMessagingSDKdelegate>

@property (nonatomic, strong) id <ConversationParamProtocol> conversationQuery;

@end

@implementation MessagingViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.accountTextField.text = @"ENTER_ACCOUNT_NUMBER";
    
    BOOL windowSwitch = [[NSUserDefaults standardUserDefaults] boolForKey:WINDOW_SWITCH];
    self.windowSwitch.on = windowSwitch;
    
    BOOL authenticationSwitch = [[NSUserDefaults standardUserDefaults] boolForKey:AUTHENTICATION_SWITCH];
    self.authenticationSwitch.on = authenticationSwitch;
    
    [LPMessaging instance].delegate = self;
    [self setSDKConfigurations];
    [[LPMessaging instance] setLoggingLevelWithLevel:LPLoggingLevelINFO];
}
    
/**
 This method sets the SDK configurations.
 For example:
 Change background color of remote user (such as Agent)
 Change background color of user (such as Consumer)
 */
- (void)setSDKConfigurations {
    LPConfig *configurations = [LPConfig defaultConfiguration];
    
    
    /**
     This method lets you enter a UIBarButton to the navigation bar (in window mode).
     When the button is pressed it will call the following delegate method: LPMessagingSDKCustomButtonTapped
     */
    UIImage *customButtonImage = [UIImage imageNamed:@"phone_icon"];
    configurations.customButtonImage = customButtonImage;
}

/**
 This method sets the user details such as first name, last name, profile image and phone number.
 */
- (void)setUserDetails {
    LPUser *user = [[LPUser alloc]
                    initWithFirstName:self.firstNameTextField.text
                    lastName:self.lastNametextField.text
                    nickName:@"my nickname"
                    uid:nil
                    profileImageURL:@"http://www.mrbreakfast.com/ucp/342_6053_ucp.jpg"
                    phoneNumber:@"000-0000000"
                    employeeID:@"1111-11111"];
    [[LPMessaging instance] setUserProfile:user brandID:self.accountTextField.text];
}

-(void)getUnreadMessageCount {
    NSString *account = self.accountTextField.text;
    if (account.length == 0) {
        NSLog(@"need account number!");
        return;
    }
    
    id<ConversationParamProtocol> conversationQueryParam = [[LPMessaging instance] getConversationBrandQuery:account campaignInfo:nil];
    [[LPMessaging instance] getUnreadMessagesCount:conversationQueryParam authenticationParams:nil completion:^(NSInteger count) {
        NSLog(@"unread message count: %ld", (long)count);
    } failure:^(NSError * _Nonnull error) {
        NSLog(@"unread message count - error: %@", error.localizedDescription);
    }];
}

#pragma mark - IBActions

- (IBAction)initSDKClicked:(id)sender {
    NSString *account = self.accountTextField.text;
    if (account.length == 0) {
        NSLog(@"need account number!");
        return;
    }
    
    NSError *error = nil;
    [[LPMessaging instance] initialize:account monitoringInitParams:nil error:&error];
    if (error) {
        NSLog(@"LPMessagingSDK Initialize Error: %@",error);
        return;
    }
}
    
    
/**
 This method shows the conversation screen. It considers different modes:
 
 Window Mode:
 - Window           - Shows the conversation screen in a new window created by the SDK. Navigation bar is included.
 - View controller  - Shows the conversation screen in a view controller of your choice.
 
 Authentication Mode:
 - Authenticated    - Conversation history is saved and shown.
 - Unauthenticated  - Conversation starts clean every time.
 
 */
- (IBAction)showConversation:(id)sender {
    NSString *account = self.accountTextField.text;
    if (account.length == 0) {
        NSLog(@"need account number!");
        return;
    }
    
    //ConversationParamProtocol
    self.conversationQuery = [[LPMessaging instance] getConversationBrandQuery:account campaignInfo:nil];
    
    //LPConversationHistoryControlParam
    LPConversationHistoryControlParam * controlParam = [[LPConversationHistoryControlParam alloc] initWithHistoryConversationsStateToDisplay: LPConversationsHistoryStateToDisplayAll
                                                                                                                 historyConversationsMaxDays:-1
                                                                                                                          historyMaxDaysType:LPConversationHistoryMaxDaysDateTypeStartConversationDate];
    //ConversationViewController
    self.conversationViewController = nil;
    
    //needed for ViewController Mode (Non-Window mode)
    if (self.windowSwitch.on == false) {
        UIStoryboard *storyBoard = [UIStoryboard storyboardWithName:@"Main" bundle:[NSBundle mainBundle]];
        self.conversationViewController = [storyBoard instantiateViewControllerWithIdentifier:@"ConversationViewController"];
        self.conversationViewController.account = account;
        self.conversationViewController.conversationQuery = self.conversationQuery;
    }
    
    //LPAuthenticationParams
    LPAuthenticationParams *authenticationParams = nil;
    if (self.authenticationSwitch.on) {
        authenticationParams = [[LPAuthenticationParams alloc] initWithAuthenticationCode:nil
                                                                                      jwt:nil
                                                                              redirectURI:nil
                                                                    certPinningPublicKeys:nil
                                                                       authenticationType:LPAuthenticationTypeAuthenticated];
    }
    
    //LPWelcomeMessageParam
    LPWelcomeMessage * welcomeMessageParam = [[LPWelcomeMessage alloc] initWithMessage:@"How may i help you today?"
                                                                             frequency:MessageFrequencyFirstTimeConversation];
//
//    LPWelcomeMessageOption * billOption = [[LPWelcomeMessageOption alloc] initWithValue:@"My latest bill statement" displayName:@"1️⃣ Bill"];
//    LPWelcomeMessageOption * orderOption = [[LPWelcomeMessageOption alloc] initWithValue:@"A recent order placed" displayName:@"2️⃣ Orders"];
//    LPWelcomeMessageOption * supportOption = [[LPWelcomeMessageOption alloc] initWithValue:@"Technical support" displayName:@"3️⃣ Support"];
//    LPWelcomeMessageOption * accountOption = [[LPWelcomeMessageOption alloc] initWithValue:@"Account information" displayName:@"4️⃣ Account"];
//    NSArray *optionsArray = [[NSArray alloc] initWithObjects:billOption, orderOption, supportOption, accountOption, nil];
//
//    NSError * error = nil;
//    [welcomeMessageParam setWithOptions:optionsArray error:&error];
//    if (error) {
//        NSLog(@"welcome message - Error: %@",error);
//    }
//
//    [welcomeMessageParam setWithNumberOfOptionsPerRow:2];
    
    //LPConversationViewParams
    LPConversationViewParams *conversationViewParams = [[LPConversationViewParams alloc] initWithConversationQuery:self.conversationQuery
                                                                                           containerViewController:self.conversationViewController
                                                                                                        isViewOnly:NO
                                                                                   conversationHistoryControlParam:controlParam
                                                                                                    welcomeMessage: welcomeMessageParam];
    
    [[LPMessaging instance] showConversation:conversationViewParams authenticationParams:authenticationParams];
    
    //needed for ViewController Mode (Non-Window mode)
    if (self.conversationViewController != nil) {
        [[self navigationController] pushViewController:self.conversationViewController animated:true];
    }
    
    [self setUserDetails];
}

- (IBAction)resignKeyboard {
    [self.view endEditing:true];
}
    
- (IBAction)windowSwitchChanged:(UISwitch *)sender {
    [[NSUserDefaults standardUserDefaults] setBool:sender.on forKey:WINDOW_SWITCH];
    [[NSUserDefaults standardUserDefaults] synchronize];
}

- (IBAction)authenticationSwitchChanged:(UISwitch *)sender {
    [[NSUserDefaults standardUserDefaults] setBool:sender.on forKey:AUTHENTICATION_SWITCH];
    [[NSUserDefaults standardUserDefaults] synchronize];
}

- (BOOL)textFieldShouldReturn:(UITextField *)textField {
    [textField resignFirstResponder];
    return NO;
}

#pragma mark - LPMessagingSDKDelegate

/**
 This delegate method is required.
 It is called when authentication process fails
 */
- (void)LPMessagingSDKAuthenticationFailed:(NSError *)error {
    NSLog(@"Error: %@",error);
}

/**
 This delegate method is required.
 It is called when the SDK version you're using is obselete and needs an update.
 */
- (void)LPMessagingSDKObseleteVersion:(NSError *)error {
    NSLog(@"Error: %@",error);
}

/**
 This delegate method is optional.
 It is called each time the SDK receives info about the agent on the other side.
 
 Example:
 You can use this data to show the agent details on your navigation bar (in view controller mode)
 */
- (void)LPMessagingSDKAgentDetails:(LPUser *)agent {
    if (agent != nil) {
        self.title = [NSString stringWithFormat:@"%@",agent.nickName];
    } else {
        self.title = @"";
    }
}

/**
 This delegate method is optional.
 It is called each time the SDK menu is opened/closed.
 */
- (void)LPMessagingSDKActionsMenuToggled:(BOOL)toggled {
    
}

/**
 This delegate method is optional.
 It is called each time the agent typing state changes.
 */
- (void)LPMessagingSDKAgentIsTypingStateChanged:(BOOL)isTyping {
    
}

/**
 This delegate method is optional.
 It is called after the customer satisfaction page is submitted with a score.
 */
- (void)LPMessagingSDKCSATScoreSubmissionDidFinish:(NSString *)accountID rating:(NSInteger)rating {
    
}

/**
 This delegate method is optional.
 If you set a custom button, this method will be called when the custom button is clicked.
 */
- (void)LPMessagingSDKCustomButtonTapped {
    
}

/**
 This delegate method is optional.
 It is called whenever an event log is received.
 */
- (void)LPMessagingSDKDidReceiveEventLog:(NSString *)eventLog {
    
}

/**
 This delegate method is optional.
 It is called when the SDK has connections issues.
 */
- (void)LPMessagingSDKHasConnectionError:(NSString *)error {
    NSLog(@"Error: %@",error);
}

/**
 This delegate method is required.
 It is called when the token which used for authentication is expired
 */
- (void)LPMessagingSDKTokenExpired:(NSString *)brandID {
    
}

/**
 This delegate method is required.
 It lets you know if there is an error with the SDK and what the error is
 */
- (void)LPMessagingSDKError:(NSError *)error {
    NSLog(@"Error: %@",error);
}

/**
 This delegate method is optional.
 It is called when a new conversation has started, from the agent or from the consumer side.
 */
- (void)LPMessagingSDKConversationStarted:(NSString *)conversationID {
    
}

/**
 This delegate method is optional.
 It is called each time connection state changed for a brand with a flag whenever connection is ready.
 Ready means that all conversations and messages were synced with the server.
 */
- (void)LPMessagingSDKConnectionStateChanged:(BOOL)isReady brandID:(NSString *)brandID {
    
}

/**
 This delegate method is optional.
 It is called when the customer satisfaction survey is dismissed after the user has submitted the survey/
 */
- (void)LPMessagingSDKConversationCSATDismissedOnSubmittion:(NSString *)conversationID {
    
}

/**
 This delegate method is optional.
 It is called when the conversation view controller removed from its container view controller or window.
 */
- (void)LPMessagingSDKConversationViewControllerDidDismiss {
    
}

/**
 This delegate method is optional.
 It is called when the user tapped on the agent’s avatar in the conversation and also in the navigation bar within window mode.
 */
- (void)LPMessagingSDKAgentAvatarTapped:(LPUser *)agent {
    
}

/**
 This delegate method is optional.
 It is called when the Conversation CSAT did load
 */
- (void)LPMessagingSDKConversationCSATDidLoad:(NSString *)conversationID {
    
}

/**
 This delegate method is optional.
 It is called when the Conversation CSAT skipped by the consumer
 */
- (void)LPMessagingSDKConversationCSATSkipped:(NSString *)conversationID {
    
}

/**
 This delegate method is optional.
 It is called when the user is opening photo sharing gallery/camera and the persmissions denied
 */
- (void)LPMessagingSDKUserDeniedPermission:(enum LPPermissionTypes)permissionType {
    
}

@end

File: Sample Apps/Swift/SampleApp-Swift/MessagingViewController.swift
//
//  ViewController.swift
//  SampleApp-Swift
//
//  Created by Nimrod Shai on 2/23/16.
//  Copyright © 2016 LivePerson. All rights reserved.
//

import UIKit
import LPMessagingSDK
import LPAMS
import LPInfra

class MessagingViewController: UIViewController {

    //MARK: - UI Properties
    @IBOutlet var accountTextField: UITextField!
    @IBOutlet var firstNameTextField: UITextField!
    @IBOutlet var lastNameTextField: UITextField!
    @IBOutlet var windowSwitch: UISwitch!
    @IBOutlet var authenticationSwitch: UISwitch!
    
    //MARK: - Properties
    private var windowSwitchValue: Bool {
        set {
            UserDefaults.standard.set(newValue, forKey: "WindowSwitch")
            UserDefaults.standard.synchronize()
        }
        get { return UserDefaults.standard.bool(forKey: "WindowSwitch") }
    }
    
    private var authenticationSwitchValue: Bool {
        set {
            UserDefaults.standard.set(newValue, forKey: "AuthenticationSwitch")
            UserDefaults.standard.synchronize()
        }
        get { return UserDefaults.standard.bool(forKey: "AuthenticationSwitch") }
    }
    
    private var conversationViewController: ConversationViewController?
    
    // Enter Your Code if using Autherization type 'Code'
    private let authenticationCode: String? = nil
    
    // Enter Your JWT if using Autherization type 'Implicit'
    private let authenticationJWT: String? = nil
    
    //MARK: - Lifecycle
    override func viewDidLoad() {
        super.viewDidLoad()
        
        // Enter Your Account Number
        self.accountTextField.text = "ENTER_ACCOUNT_NUMBER"
        
        self.windowSwitch.isOn = windowSwitchValue
        self.authenticationSwitch.isOn = authenticationSwitchValue
        
        LPMessagingSDK.instance.delegate = self
        self.setSDKConfigurations()
        LPInfraFacade.setLoggingLevel(level: .INFO)
    }

    //MARK: - IBActions
    @IBAction func resignKeyboard() {
        self.view.endEditing(true)
    }
    
    @IBAction func windowSwitchChanged(_ sender: UISwitch) {
        windowSwitchValue = sender.isOn
    }
    
    @IBAction func authenticationSwitchChanged(_ sender: UISwitch) {
        authenticationSwitchValue = sender.isOn
    }
    
    @IBAction func initSDKsClicked(_ sender: Any) {
        defer { self.view.endEditing(true) }
        
        guard let accountNumber = self.accountTextField.text, !accountNumber.isEmpty else {
            print("missing account number!")
            return
        }
        
        initLPSDKwith(accountNumber: accountNumber)
    }

    @IBAction func showConversation() {
        defer { self.view.endEditing(true) }
        
        guard let accountNumber = self.accountTextField.text, !accountNumber.isEmpty else {
            print("missing account number!")
            return
        }
        
        //Window Mode
        if windowSwitchValue {
            self.conversationViewController = nil
        } else {
            //for ViewController Mode ONLY
            let storyboard = UIStoryboard(name: "Main", bundle: nil)
            let controller = storyboard.instantiateViewController(withIdentifier: "ConversationViewController")
            guard let convoViewController = controller as? ConversationViewController else { return }
            self.conversationViewController = convoViewController
        }
 
        showConversationFor(accountNumber: accountNumber, authenticatedMode: authenticationSwitchValue)
        
        //do not forget to push the controller (for ViewController Mode ONLY)
        if self.conversationViewController != nil {
            self.navigationController?.pushViewController(self.conversationViewController!, animated: true)
        }
    }
    
    @IBAction func logoutClicked(_ sender: Any) {
        logoutLPSDK()
    }
}

// MARK: - LPMessagingSDK Helpers
extension MessagingViewController {
    /**
     This method sets the SDK configurations.
     
     For example:
         Change background color of remote user (such as Agent)
         Change background color of user (such as Consumer)
     
    for more information on `defaultConfiguration` see: https://developers.liveperson.com/mobile-app-messaging-sdk-for-ios-customization-and-branding-customizing-the-sdk.html
     */
    private func setSDKConfigurations() {
        let configurations = LPConfig.defaultConfiguration
        
        /* the below  lets you enter a UIBarButton to the navigation bar (in window mode).
         When the button is pressed it will call the following delegate method: LPMessagingSDKCustomButtonTapped */
        configurations.customButtonImage = UIImage(named: "phone_icon")
    }
    
    private func getUnreadMessageCount() {
        guard let accountNumber = self.accountTextField.text, !accountNumber.isEmpty else {
            print("missing account number!")
            return
        }
        
        //NOTE: After SDK version 4.1.0
//        LPMessagingSDK.getUnreadMessagesCount(brandID: accountNumber, completion: { (count) in
//            print("unread message count: \(count)")
//        }) { (error) in
//            print("unread message count - error: \(error.localizedDescription)")
//        }
        
        //NOTE: Before SDK version 4.1.0
        let conversationQuery = LPMessagingSDK.instance.getConversationBrandQuery(accountNumber)
        LPMessagingSDK.getUnreadMessagesCount(conversationQuery, completion: { (count) in
            print("unread message count: \(count)")
        }) { (error) in
            print("unread message count - error: \(error.localizedDescription)")
        }
    }
    
    /**
     This method initialize the messaging SDK
     
     for more information on `initialize` see:
         https://developers.liveperson.com/mobile-app-messaging-sdk-for-ios-sdk-apis-messaging-api.html#initialize
     */
    private func initLPSDKwith(accountNumber: String){
        do {
            try LPMessagingSDK.instance.initialize(accountNumber)
        } catch let error as NSError {
            print("initialize error: \(error)")
        }
    }
    
    /**
     This method shows the conversation screen. It considers different modes:
     
     Window Mode:
     - Window           - Shows the conversation screen in a new window created by the SDK. Navigation bar is included.
     - View controller  - Shows the conversation screen in a view controller of your choice.
     
     Authentication Mode:
     - Authenticated    - Conversation history is saved and shown.
     - Unauthenticated  - Conversation starts clean every time.
     
     for more information on `showconversation` see:
         https://developers.liveperson.com/mobile-app-messaging-sdk-for-ios-sdk-apis-messaging-api.html#showconversation
     */
    private func showConversationFor(accountNumber: String, authenticatedMode: Bool) {
        //ConversationParamProtocol
        let conversationQuery = LPMessagingSDK.instance.getConversationBrandQuery(accountNumber)
        
        //LPConversationHistoryControlParam
        let controlParam = LPConversationHistoryControlParam(historyConversationsStateToDisplay: .all,
                                                             historyConversationsMaxDays: -1,
                                                             historyMaxDaysType: .startConversationDate)

        //LPAuthenticationParams
        var authenticationParams: LPAuthenticationParams?
        if authenticatedMode {
            authenticationParams = LPAuthenticationParams(authenticationCode: authenticationCode,
                                                          jwt: authenticationJWT,
                                                          redirectURI: nil,
                                                          certPinningPublicKeys: nil,
                                                          authenticationType: .authenticated)
        }
        
        // update Account number and ConversationQuery (for ViewController Mode ONLY)
        if self.conversationViewController != nil {
            self.conversationViewController?.account = accountNumber
            self.conversationViewController?.conversationQuery = conversationQuery
        }
        
        //LPWelcomeMessageParam
        let welcomeMessageParam = LPWelcomeMessage(message: LanguagesManager.localizedStringForKey(LPLanguagesKeys.hiMessage), frequency: .FirstTimeConversation)
        
//        let welcomeMessageOptions = [
//            LPWelcomeMessageOption(value: "My latest bill statement", displayName: "1️⃣ Bill"),
//            LPWelcomeMessageOption(value: "A recent order placed", displayName: "2️⃣ Orders"),
//            LPWelcomeMessageOption(value: "Technical support", displayName: "3️⃣ Support"),
//            LPWelcomeMessageOption(value: "Account information", displayName: "4️⃣ Account")
//        ]
//
//        do {
//            try welcomeMessageParam.set(options: welcomeMessageOptions)
//        }
//        catch {
//            print("cannot set welcome message options | error: \(error.localizedDescription)")
//        }
//
//        welcomeMessageParam.set(NumberOfOptionsPerRow: 2)
        
        //LPConversationViewParams
        let conversationViewParams = LPConversationViewParams(conversationQuery: conversationQuery,
                                                              containerViewController: self.conversationViewController,
                                                              isViewOnly: false,
                                                              conversationHistoryControlParam: controlParam,
                                                              welcomeMessage: welcomeMessageParam)
        
        LPMessagingSDK.instance.showConversation(conversationViewParams, authenticationParams: authenticationParams)

        self.setUserDetails()
    }
    
    /**
     This method sets the user details such as first name, last name, profile image and phone number.
     
     for more info on `setUserProfile` see:
         https://developers.liveperson.com/mobile-app-messaging-sdk-for-ios-sdk-apis-messaging-api.html#setuserprofile
     */
    private func setUserDetails() {
        let user = LPUser(firstName: self.firstNameTextField.text!,
                          lastName: self.lastNameTextField.text!,
                          nickName: "my nick name",
                          uid: nil,
                          profileImageURL: "http://www.mrbreakfast.com/ucp/342_6053_ucp.jpg",
                          phoneNumber: nil,
                          employeeID: "1111-1111")
        LPMessagingSDK.instance.setUserProfile(user, brandID: self.accountTextField.text!)
    }
    
    /**
     This method logouts from Monitoring and Messaging SDKs - all the data will be cleared
     
     for more info on `logout` see:
         https://developers.liveperson.com/mobile-app-messaging-sdk-for-ios-methods-logout.html
     */
    private func logoutLPSDK() {
        LPMessagingSDK.instance.logout(completion: {
            print("successfully logout from MessagingSDK")
        }) { (error) in
            print("failed to logout from MessagingSDK - error: \(error.localizedDescription)")
        }
    }
}

//MARK: - LPMessagingSDKdelegate

/**
 for more info on `LPMessagingSDKdelegate` see:
     https://developers.liveperson.com/mobile-app-messaging-sdk-for-ios-sdk-apis-callbacks-index.html#lpmessagingsdkdelegate
 */
extension MessagingViewController: LPMessagingSDKdelegate {
    
    /**
    This delegate method is required.
    It is called when authentication process fails
    */
    func LPMessagingSDKAuthenticationFailed(_ error: NSError) {
        NSLog("Error: \(error)")
    }
    
    /**
    This delegate method is required.
    It is called when the SDK version you're using is obselete and needs an update.
    */
    func LPMessagingSDKObseleteVersion(_ error: NSError) {
        NSLog("Error: \(error)")
    }
    
    /**
    This delegate method is optional.
    It is called each time the SDK receives info about the agent on the other side.
    
    Example:
    You can use this data to show the agent details on your navigation bar (in view controller mode)
    */
    func LPMessagingSDKAgentDetails(_ agent: LPUser?) {
        guard self.conversationViewController != nil else { return }

        let name: String = agent?.nickName ?? ""
        self.conversationViewController?.title = name
    }
    
    /**
    This delegate method is optional.
    It is called each time the SDK menu is opened/closed.
    */
    func LPMessagingSDKActionsMenuToggled(_ toggled: Bool) {
        
    }
    
    /**
    This delegate method is optional.
    It is called each time the agent typing state changes.
    */
    func LPMessagingSDKAgentIsTypingStateChanged(_ isTyping: Bool) {
        
    }
    
    /**
    This delegate method is optional.
    It is called after the customer satisfaction page is submitted with a score.
    */
    func LPMessagingSDKCSATScoreSubmissionDidFinish(_ accountID: String, rating: Int) {
    
    }
    
    /**
    This delegate method is optional.
    If you set a custom button, this method will be called when the custom button is clicked.
    */
    func LPMessagingSDKCustomButtonTapped() {
        
    }
    
    /**
    This delegate method is optional.
    It is called whenever an event log is received.
    */
    func LPMessagingSDKDidReceiveEventLog(_ eventLog: String) {
        
    }
    
    /**
    This delegate method is optional.
    It is called when the SDK has connections issues.
    */
    func LPMessagingSDKHasConnectionError(_ error: String?) {
        NSLog("Error: \(String(describing: error))")
    }

    /**
     This delegate method is required.
     It is called when the token which used for authentication is expired
     */
    func LPMessagingSDKTokenExpired(_ brandID: String) {
        NSLog("LPMessagingSDKTokenExpired")
    }
    
    /**
     This delegate method is required.
     It lets you know if there is an error with the sdk and what this error is
     */
    func LPMessagingSDKError(_ error: NSError) {
        NSLog("Error: \(error)")
    }
    
    /**
     This delegate method is optional.
     It is called when the conversation view controller removed from its container view controller or window.
     */
    func LPMessagingSDKConversationViewControllerDidDismiss() {
        
    }
    
    /**
     This delegate method is optional.
     It is called when a new conversation has started, from the agent or from the consumer side.
     */
    func LPMessagingSDKConversationStarted(_ conversationID: String?) {
        
    }
    
    /**
     This delegate method is optional.
     It is called when a conversation has ended, from the agent or from the consumer side.
     */
    func LPMessagingSDKConversationEnded(_ conversationID: String?) {
        
    }
    
    /**
     This delegate method is optional.
     It is called when the customer satisfaction survey is dismissed after the user has submitted the survey/
     */
    func LPMessagingSDKConversationCSATDismissedOnSubmittion(_ conversationID: String?) {
        
    }
    
    /**
     This delegate method is optional.
     It is called each time connection state changed for a brand with a flag whenever connection is ready.
     Ready means that all conversations and messages were synced with the server.
     */
    func LPMessagingSDKConnectionStateChanged(_ isReady: Bool, brandID: String) {
        
    }
    
    /**
     This delegate method is optional.
     It is called when the user tapped on the agent’s avatar in the conversation and also in the navigation bar within window mode.
     */
    func LPMessagingSDKAgentAvatarTapped(_ agent: LPUser?) {
        
    }
    
    /**
     This delegate method is optional.
     It is called when the Conversation CSAT did load
    */
    func LPMessagingSDKConversationCSATDidLoad(_ conversationID: String?) {
        
    }
    
    /**
     This delegate method is optional.
     It is called when the Conversation CSAT skipped by the consumer
     */
    func LPMessagingSDKConversationCSATSkipped(_ conversationID: String?) {
        
    }
    
    /**
     This delegate method is optional.
     It is called when the user is opening photo sharing gallery/camera and the persmissions denied
    */
    func LPMessagingSDKUserDeniedPermission(_ permissionType: LPPermissionTypes) {
        
    }
}
